pkg/controller/endpointslicemirroring/utils.go

```diff
@@ -0,0 +1,246 @@
+	}
+	return &addressType
+}
+
+// endpointsEqualBeyondHash returns true if endpoints have equal attributes
+// but excludes equality checks that would have already been covered with
+// endpoint hashing (see hashEndpoint func for more info).
+func endpointsEqualBeyondHash(ep1, ep2 *discovery.Endpoint) bool {
+	if !apiequality.Semantic.DeepEqual(ep1.Topology, ep2.Topology) {
+		return false
+	}
+
+	if boolPtrChanged(ep1.Conditions.Ready, ep2.Conditions.Ready) {
+		return false
+	}
+
+	if objectRefPtrChanged(ep1.TargetRef, ep2.TargetRef) {
+		return false
+	}
+
+	return true
+}
+
+// newEndpointSlice returns an EndpointSlice generated from an Endpoints
+// resource, ports, and address type.
+func newEndpointSlice(endpoints *corev1.Endpoints, ports []discovery.EndpointPort, addrType discovery.AddressType) *discovery.EndpointSlice {
+	gvk := schema.GroupVersionKind{Version: "v1", Kind: "Endpoints"}
+	ownerRef := metav1.NewControllerRef(endpoints, gvk)
+	return &discovery.EndpointSlice{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: map[string]string{
+				discovery.LabelServiceName: endpoints.Name,
+				discovery.LabelManagedBy:   controllerName,
+			},
+			GenerateName:    getEndpointSlicePrefix(endpoints.Name),
+			OwnerReferences: []metav1.OwnerReference{*ownerRef},
+			Namespace:       endpoints.Namespace,
+		},
+		Ports:       ports,
+		AddressType: addrType,
+		Endpoints:   []discovery.Endpoint{},
+	}
+}
+
+// getEndpointSlicePrefix returns a suitable prefix for an EndpointSlice name.
+func getEndpointSlicePrefix(serviceName string) string {
+	// use the dash (if the name isn't too long) to make the name a bit prettier.
+	prefix := fmt.Sprintf("%s-", serviceName)
+	if len(validation.ValidateEndpointSliceName(prefix, true)) != 0 {
+		prefix = serviceName
+	}
+	return prefix
+}
+
+// addressToEndpoint converts an address from an Endpoints resource to an
+// EndpointSlice endpoint.
+func addressToEndpoint(address corev1.EndpointAddress, ready bool) *discovery.Endpoint {
+	endpoint := &discovery.Endpoint{
+		Addresses: []string{address.IP},
+		Conditions: discovery.EndpointConditions{
+			Ready: &ready,
+		},
+		TargetRef: address.TargetRef,
+	}
+
+	if address.NodeName != nil {
+		endpoint.Topology = map[string]string{
+			"kubernetes.io/hostname": *address.NodeName,
+		}
+	}
+	if address.Hostname != "" {
+		endpoint.Hostname = &address.Hostname
+	}
+
+	return endpoint
+}
+
+// epPortsToEpsPorts converts ports from an Endpoints resource to ports for an
+// EndpointSlice resource.
+func epPortsToEpsPorts(epPorts []corev1.EndpointPort) []discovery.EndpointPort {
+	epsPorts := []discovery.EndpointPort{}
+	for _, epPort := range epPorts {
+		epp := epPort.DeepCopy()
+		epsPorts = append(epsPorts, discovery.EndpointPort{
+			Name:        &epp.Name,
+			Port:        &epp.Port,
+			Protocol:    &epp.Protocol,
+			AppProtocol: epp.AppProtocol,
+		})
+	}
+	return epsPorts
+}
+
+// boolPtrChanged returns true if a set of bool pointers have different values.
```

Comment:
boolPtrEqual would be the more common way to name this

Reply:
Updated, thanks!